### PR TITLE
Global Illumination improvements

### PR DESCRIPTION
**Pull request by jwapobie**
A collection of improvements to the GI system which makes its effect much more visible to the player. These changes are needed to increase GI light contribution without noticeable artifacts.

- Replace vanilla ambient light with GI raycasted light
- Use cave factor to reduce ambient light leakage underground
- Add setting to mix in vanilla ambient light back in for gameplay purposes
- Increase GI contributions from sunlight and blocklight
- Sample normal at raycast hits for sunlight intensity
- Add GI contributions to raycast contribution to simulate bounce light
- Fix GI normals in corners, which fixes most GI light leakage
- Bypass sdf < 0.7 requirement with 1/37 chance, so non-voxelized entities and blocks can still receive accurate ambient light
- Use max of all DF values sampled during normal calculation for sdf > 0.1 requirement, to fix edge case with slabs in recessed walls no longer receiving ambient light

### CODE REVIEW

File: shaders/program/shadowcomp1.glsl [hunk 1]
```diff
@@ -410,8 +415,19 @@ void main() {
 
         if (insideFrustrum) {
             float thisDFval = getDistanceField(vxPos);
-            if (thisDFval < 0.7) {
-                if (thisDFval > 0.1) {
+            if (thisDFval < 0.7 || nextUint() % 37 == 0) {
```
Comment (gri573) [inline shaders/program/shadowcomp1.glsl:418]: I think this (randomly picking threads to do the calculations anyway) is probably very bad for coherency, meaning that most warps will have approx. one thread executing the calculations, so the majority of warps will be mostly stalled, and only a couple will be able to skip this part. I'll figure out a better way, such as marking interesting locations for GI that should also be calculated

File: shaders/program/shadowcomp1.glsl [hunk 2]
```diff
@@ -422,30 +438,37 @@ void main() {
...
-                    normal = normalize(normal);
                     vxPos -= min(0.3, thisDFval - 0.1) * normal;
                     for (int sampleNum = 0; sampleNum < GI_SAMPLE_COUNT; sampleNum++) {
                         vec3 dir = randomSphereSample();
                         if (dot(dir, normal) < 0.0) dir = -dir;
                         float ndotl = dot(dir, normal);
                         vec3 hitPos = rayTrace(vxPos, LIGHT_TRACE_LENGTH * dir, dither);
-                        vec3 hitCol = vec3(0);
+                        #ifdef GL_CAVE_FACTOR
+                            vec3 hitCol = ambientColor * clamp(dir.y + 1.6, 0.6, 1) * (1-GetCaveFactor(cameraPosition.y + vxPos.y)) / GI_STRENGTH;
```
Comment (gri573) [inline shaders/program/shadowcomp1.glsl:448]: I think I store skylight somewhere in the voxel data, could probably use that. will check where it is and try if it looks alright

File: shaders/program/shadowcomp1.glsl [hunk 3]
```diff
@@ -410,8 +415,19 @@ void main() {
...
             float thisDFval = getDistanceField(vxPos);
-            if (thisDFval < 0.7) {
-                if (thisDFval > 0.1) {
+            if (thisDFval < 0.7 || nextUint() % 37 == 0) {
+                bool isOccluded = (imageLoad(occupancyVolume, coords).r & 1) != 0;
+                float maxDFVal = thisDFval;
+                for (int k = 0; k < 3; k++) {
+                    float dplus = getDistanceField(vxPos + mat3(0.5)[k]);
+                    float dminus = getDistanceField(vxPos - mat3(0.5)[k]);
+                    if (isOccluded) dplus = -dplus;
```
Comment (gri573) [inline shaders/program/shadowcomp1.glsl:424]: could do a single if check, would be closer to my code style but probably doesn't matter much

File: shaders/program/shadowcomp1.glsl [hunk 4]
```diff
@@ -410,8 +415,19 @@ void main() {
...
+                    if (isOccluded) dminus = -dminus;
```
Comment (gri573) [inline shaders/program/shadowcomp1.glsl:425]: could do a single if check, would be closer to my coding style, but probably doesn't matter much

File: shaders/program/shadowcomp1.glsl (hunk 2 above)
Comment (gri573) [reply shaders/program/shadowcomp1.glsl:448]: Never mind that, it would look bad

File: shaders/program/shadowcomp1.glsl (hunk 4 above)
Comment (gri573) [reply shaders/program/shadowcomp1.glsl:425]: one could vectorize it by creating a vec2 offsetDist = vec2(dplus, dminus). That is a micro-optimization though

File: shaders/program/shadowcomp1.glsl (hunk 1 above)
Comment (jwapobie) [reply shaders/program/shadowcomp1.glsl:418]: Will leave it up to you how to address this, its a bit beyond my current knowledge. I also find that these random air samples look pretty noisy, so they could have increased influence from neighbors

File: shaders/program/shadowcomp1.glsl (hunk 2 above)
Comment (jwapobie) [reply shaders/program/shadowcomp1.glsl:448]: IMO Caves look decent rn, but xlarge builds still have light coming from nowhere. Maybe could be improved with skylight, but dont want to affect other situations too much

File: shaders/program/shadowcomp1.glsl (hunk 4 above)
Comment (jwapobie) [reply shaders/program/shadowcomp1.glsl:425]: I'll combine into a single if check

**Pull request merged**